lib/recursive-methods.rb

```diff
@@ -1,49 +1,89 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(nlogn)
+# Space complexity: O(nlogn)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  if n < 0
+    raise ArgumentError, "Error!!"
+  end
+
+  if n == 0 || n == 1
+    return 1
+  else
+    return n * factorial(n - 1)
+  end
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
```

Comment:
This is O(n^2) since `s[0...-1]` creates a new string of length n-1 each iteration.  Both for space and time.